src/nubank/auth/mod.rs

```diff
@@ -0,0 +1,182 @@
+
+    match urls.get(&name) {
+        Some(v) => Ok(v.to_string().replace("\"", "")),
+        _ => return Err(AuthError::UrlNameNotFoundInDiscoveryJson(name)),
+    }
+}
+
+fn get_identity(path: String) -> Result<reqwest::Identity, AuthError> {
+    let mut buf = Vec::new();
+    let _ = File::open(path)
+        .map_err(|x| AuthError::CannotReadCertificate(x))?
+        .read_to_end(&mut buf);
+    let id = reqwest::Identity::from_pkcs12_der(&buf, "")
+        .map_err(|x| AuthError::CannotInterpretCertificate(x))?;
+    Ok(id)
+}
+
+fn build_client(id: reqwest::Identity) -> Result<reqwest::Client, AuthError> {
+    reqwest::Client::builder()
+        .identity(id)
+        .build()
+        .map_err(|x| AuthError::CannotBuildClient(x))
+}
+
+fn build_payload(cpf: String, password: String) -> String {
+    json!(
+        {
+            "grant_type": "password",
+            "client_id": "legacy_client_id",
+            "client_secret": "legacy_client_secret",
+            "login": cpf,
+            "password": password
+        }
+    )
+    .to_string()
+}
+
+async fn make_auth_request(
+    client: reqwest::Client,
+    url: String,
+    payload: String,
+) -> Result<String, AuthError> {
+    let result = client
+        .post(url)
+        .apply_default_header()
+        .body(payload)
+        .send()
+        .await
+        .map_err(|x| AuthError::AuthRequestFailed(x))?
+        .text()
+        .await
+        .map_err(|x| AuthError::AuthRequestDecodingFailed(x))?;
+
+    Ok(result)
+}
+fn read_request_output(result: String) -> Result<AuthDataDTO, AuthError> {
+    let obj = serde_json::from_str::<AuthDataDTO>(&*result)
+        .map_err(|x| AuthError::AuthJsonConversionFailed(x))?;
+
+    Ok(obj)
+}
+
+fn build_auth_data_obj(auth_data_dto: AuthDataDTO) -> Result<AuthData, AuthError> {
+    match (
+        first_or(
+            auth_data_dto.links.get("events"),
+            auth_data_dto.links.get("magnitude"),
+        ),
+        auth_data_dto.links.get("bills_summary"),
+        auth_data_dto.links.get("customer"),
+        auth_data_dto.links.get("ghostflame"),
+        auth_data_dto.links.get("revoke_token"),
+    ) {
+        (
+            Some(feed_url),
+            Some(bills_url),
+            Some(customer_url),
+            Some(query_url),
+            Some(revoke_token_url),
+        ) => Ok(AuthData {
+            access_token: auth_data_dto.access_token,
+            feed_url: feed_url.clone().href,
+            bills_url: bills_url.clone().href,
+            customer_url: customer_url.clone().href,
+            query_url: query_url.clone().href,
+            revoke_token_url: revoke_token_url.clone().href,
+        }),
+        _ => return Err(AuthError::RequiredFieldsNotFoundInAuthJson(auth_data_dto)),
+    }
+}
+
+fn first_or<T>(a: Option<T>, b: Option<T>) -> Option<T> {
+    match a {
+        Some(v) => Some(v),
+        None => b,
+    }
+}
```

Comment:
please extract this one into a proper utils module

Reply:
Done.